CONFIGURATION.md
# Configuration

Document Merge Service is a [12factor app](https://12factor.net/) which means
that configuration is stored in environment variables. Different environment
variable types are explained at
[django-environ](https://github.com/joke2k/django-environ#supported-types).

## Common

- `SECRET_KEY`: A secret key used for cryptography. This needs to be a random
  string of a certain length. See
  [more](https://docs.djangoproject.com/en/2.1/ref/settings/#std:setting-SECRET_KEY).
- `ALLOWED_HOSTS`: A list of hosts/domains your service will be served from. See
  [more](https://docs.djangoproject.com/en/2.1/ref/settings/#allowed-hosts).

## Database

Per default [Sqlite3](https://sqlite.org/) is used as database for simple
deployment and stored at `/var/lib/document-merge-service/data/sqlite3.db`.
Create a volume to make it persistent.

To scale the service a different database storage is needed. Any database
supported by
[Django](https://docs.djangoproject.com/en/2.1/ref/settings/#std:setting-DATABASE-ENGINE)
can be used.

- `DATABASE_ENGINE`: Database backend to use.
- `DATABASE_HOST`: Host to use when connecting to database
- `DATABASE_PORT`: Port to use when connecting to database
- `DATABASE_NAME`: Name of database to use
- `DATABASE_USER`: Username to use when connecting to the database
- `DATABASE_PASSWORD`: Password to use when connecting to database

## Unoconv

- `UNOCONV_ALLOWED_TYPES`: List of types allowed to convert to. See
  `unoconv --show` (default: ['pdf'])
- `UNOCONV_PYTHON`: String, defaults to "/usr/bin/python3.5"
- `UNOCONV_PATH`: String, defaults to "/usr/bin/unoconv"

## python-docx-template

- `DOCXTEMPLATE_JINJA_EXTENSIONS`: list of
  [jinja2 extensions](http://jinja.pocoo.org/docs/2.10/extensions/) to load

In python-docx-template following additional custom filters are implemented:

- multiline(value) - wraps the value in a
  [Listing](https://docxtpl.readthedocs.io/en/latest/#escaping-newline-new-paragraph-listing)
  for multiline support
- datetimeformat(value, format, locale)
- dateformat(value, format, locale)
- timeformat(value, format, locale)
- getwithdefault(value, default) - converts None to empty string (or provided
  default value) or leaves strings as is
- emptystring(value) - converts None to empty string or leaves strings as is
  (deprecated in favor of getwithdefault)
- image(width, height, keep_aspect_ratio) - Creates an
  [inline image](https://docxtpl.readthedocs.io/en/latest/) from provided file
  with the same name. `width` and `height` are optional and represent
  millimetres. If `keep_aspect_ratio` is `True` the image will be scaled keeping
  it's original aspect ratio and the width/height parameters become a size limit
  instead. `keep_aspect_ration` has no effect if `width` and `height` are not
  given.

For formatting use babel and its uniode compatible
[format](http://babel.pocoo.org/en/latest/dates.html#date-fields).

## Authentication / Authorization

By default, no authentication is needed. To protect the API, integrate it with
your [IAM](https://en.wikipedia.org/wiki/Identity_management) supporting Open ID
Connect. If not available, you might consider using
[Keycloak](https://www.keycloak.org/).

- `REQUIRE_AUTHENTICATION`: Force authentication to be required (default: False)
- `OIDC_USERINFO_ENDPOINT`: Url of userinfo endpoint as
  [described](https://openid.net/specs/openid-connect-core-1_0.html#UserInfo)
- `OIDC_VERIFY_SSL`: Verify ssl certificate of oidc userinfo endpoint (default:
  True)
- `OIDC_GROUPS_CLAIM`: Name of claim to be used to define group membership
  (default: document_merge_service_groups)
- `OIDC_USERNAME_CLAIM`: Name of claim to be used to define user (default: sub)
- `OIDC_BEARER_TOKEN_REVALIDATION_TIME`: Time in seconds before bearer token
  validity is verified again. For best security token is validated on each
  request per default. It might be helpful though in case of slow Open ID
  Connect provider to cache it. It uses [cache](#cache) mechanism for memorizing
  userinfo result. Number has to be lower than access token expiration time.
  (default: 0)

## Permissions / Visibilities

Document Merge Service uses
[dgap](https://github.com/adfinis/django-generic-api-permissions) to handle
permissions and visibilities. It can be configured using the following
environment variables:

- `DMS_VISIBILITY_CLASSES`: List of classes that handle
  [dgap visibilities](https://github.com/adfinis/django-generic-api-permissions#visibilities)
- `DMS_PERMISSION_CLASSES`: List of classes that handle
  [dgap permissions](https://github.com/adfinis/django-generic-api-permissions#permissions)
- `EXTENSIONS_ARGUMENTS`: Custom arguments from the app to be used in the
  visibility and permission classes. This is expected to be a `dict`, e.g.
  `EXTENSIONS_ARGUMENTS=foo=bar` could then be used in the extension classes as
  `settings.EXTENSIONS_ARGUMENTS["foo"]`

## Cache

- `CACHE_BACKEND`:
  [cache backend](https://docs.djangoproject.com/en/4.2/ref/settings/#backend)
  to use (default: django.core.cache.backends.locmem.LocMemCache)
- `CACHE_LOCATION`:
  [location](https://docs.djangoproject.com/en/4.2/ref/settings/#location) of
  cache to use
- `CACHE_OPTIONS`:
  [options](https://docs.djangoproject.com/en/4.2/ref/settings/#options) for
  cache library

## CORS

- `CORS_ORIGIN_ALLOW_ALL`:
  [allow all](https://github.com/ottoyiu/django-cors-headers#cors_origin_allow_all)
- `CORS_ORIGIN_REGEX_WHITELIST`: List of
  [whitelist regexes](https://github.com/ottoyiu/django-cors-headers#cors_origin_regex_whitelist)
  defaults to "^(https?://)?127\.0\.0\.1:\d{4}$"

Users of nginx/apache must ensure to have matching CORS configurations.

## Pagination

- `PAGINATION_ENABLED`: whether the pagination is enabled (default: `True`)
- `PAGINATION_DEFAULT_PAGE_SIZE`: the default page size if no query param
  (`page_size`) is given (default: `100`)
- `PAGINATION_MAX_PAGE_SIZE`: the max value of the page size query param
  (`page_size`) (default: `1000`)

## Email

- `SERVER_EMAIL`: the email address that error messages come from
- `DEFAULT_FROM_EMAIL`: default email address to use for various automated
  correspondence. This doesn’t include error messages sent to `ADMINS`.
- `EMAIL_HOST`: the host to use for sending email (default: `localhost`)
- `EMAIL_PORT`: port to use for the SMTP server (default: `25`)
- `EMAIL_HOST_USER`: username for the SMTP server(default: "")
- `EMAIL_HOST_PASSWORD`: password for the SMTP server user (default: "")
- `EMAIL_USE_TLS`: whether to use an implicit TLS (secure) connection when
  talking to the SMTP server (default: `False`)

If either `EMAIL_HOST_USER` or `EMAIL_HOST_PASSWORD` is empty, Django won't
attempt authentication.

## Email error handler

- `ENABLE_ADMIN_EMAIL_LOGGING`: enable Django to send email to admins on errors
  (default: `False`)
- `ADMINS`: list of people who will get code error notifications. Items in the
  list should follow this example:
  `Test Example <test@example.com>,Test2 <test2@example.com>`

## Sentry

- `SENTRY_DSN`: identifier (data source name) for where to send events to. If no
  value is provided, sentry won't be activated (default: "")
- `SENTRY_ENVIRONMENT`: which app environment sent an event to sentry (default:
  `development`)
- `SENTRY_TRACES_SAMPLE_RATE`: percentage chance a given transaction will be
  sent to Sentry (default: `1.0`)
- `SENTRY_SEND_DEFAULT_PII`: enable send PII data that associates users to
  errors (default: `True`)

## Template storage

- `FILE_STORAGE`: Django file storage backend (default:
  `django.core.files.storage.FileSystemStorage`)
- `MEDIA_ROOT`: Absolute filesystem path to the directory that will hold
  user-uploaded files. (default: "")
- `MEDIA_URL`: URL that handles the media served from MEDIA_ROOT, used for
  managing stored files. When using buckets this needs to be changed. (default:
  `api/v1/template/`)

### [django-storages](https://django-storages.readthedocs.io/en/latest/backends/amazon-S3.html) S3 settings

Refer to for example
[Digital Ocean](https://django-storages.readthedocs.io/en/latest/backends/s3_compatible/digital-ocean-spaces.html)
configuration if using a S3 compatible storage which isn't AWS.

Required to use S3 storage:

- `DMS_S3_ACCESS_KEY_ID`: AWS access key id
- `DMS_S3_SECRET_ACCESS_KEY`: AWS secret access key
- `DMS_STORAGE_BUCKET_NAME`: Storage bucket name

Optional:

- `DMS_S3_ENDPOINT_URL`: Custom S3 URL to use when connecting to S3, including
  scheme. (default: "")
- `DMS_S3_REGION_NAME`: Region of the storage (default: "")
- `DMS_LOCATION`: A path prefix that will be prepended to all uploads (default:
  "")
- `DMS_S3_FILE_OVERWRITE`: If `True` Files with the same name will overwrite
  each other. Otherwise extra characters are appended. (default: `False`)
- `DMS_S3_SIGNATURE_VERSION`: S3 signature version to use (default: `s2`)
- `DMS_S3_USE_SSL`: Whether or not to use SSL when connecting to S3 (default:
  `True`)
- `DMS_S3_VERIFY`: Whether or not to verify the connection to S3. Can be set to
  False to not verify SSL/TLS certificates. (default: `None`)
- `DMS_ENABLE_AT_REST_ENCRYPTION`: Whether to use SSEC to encrypt files uploaded
  to S3 (default: `False`)
- `DMS_S3_STORAGE_SSEC_SECRET`: Secret key for SSEC encryption, has to be 32
  bytes long (default: `xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx`)

## Gunicorn

Document Merge Service uses [Gunicorn](https://gunicorn.org/) as its app server.
A reasonable configuration is included in the docker image and can be customized
using the following environment variables:

- `GUNICORN_HOST`: IP address to bind gunicorn to (default `0.0.0.0`)
- `GUNICORN_PORT`: Port to bind gunicorn to (default `8000`)
- `GUNICORN_WORKERS`: Number of workers for handling requests (default `8`)
- `GUNICORN_TIMEOUT`: Number of seconds until worker processing a request is
  killed and restarted (default `60`)
- `GUNICORN_LIMIT_REQUEST_LINE`: Maximum size of HTTP request line in bytes
  (default `8190`)
- `GUNICORN_MAX_REQUESTS`: The maximum number of requests a worker will process
  before restarting (default `1000`)
- `GUNICORN_MAX_REQUESTS_JITTER`: The maximum jitter to add to the max_requests
  setting (default `100`)

If you want to further customize your configuration, you may do that by
overwriting the file
[`document_merge_service/gunicorn.py`](document_merge_service/gunicorn.py) and
add your own custom settings:

```dockerfile
FROM ghcr.io/adfinis/document-merge-service:latest

COPY my_gunicorn_config.py /app/document_merge_service/gunicorn.py
```

Such a configuration file might look like this:

```python
# my_gunicorn_config.py

wsgi_app = "document_merge_service.wsgi:application" # must not be changed
bind = "0.0.0.0:80"
workers = 16
timeout = 120
loglevel = "debug"
```

For more information on how to customize Gunicorn, please refer to
[the official documentation](https://docs.gunicorn.org/en/latest/settings.html).


CHANGELOG.md
# Changelog

## 9.1.1 (17 February 2026)

### Fix

* Force write to disk before call to unoconv ([`785e1e8`](https://github.com/adfinis/document-merge-service/commit/785e1e8f73b025472d0c80ae60bec5286305f653))

## 9.1.0 (13 February 2026)

### Feature

* **gunicorn:** Add max-requets and max-requests-jitter env support ([`c90939a`](https://github.com/adfinis/document-merge-service/commit/c90939a4cce24e1f767c8edcf7a29290861d0aa8))

### Fix

* Use our fork of xltpl ([#1098](https://github.com/adfinis/document-merge-service/issues/1098)) ([`df78a42`](https://github.com/adfinis/document-merge-service/commit/df78a42c2436a4a3eddf571bb979668826047ed4))

## 9.0.0 (19 January 2026)

### Feature

* **docker:** Update to python v3.14 ([`28df1b3`](https://github.com/adfinis/document-merge-service/commit/28df1b3e5230120c7dbda7cabf0256cb0ad9e5a4))

### Fix

* **deps:** Update dependencies ([`90e7202`](https://github.com/adfinis/document-merge-service/commit/90e72025076601d60aea27f3deac559a0c3e5770))

### Breaking

* This removes support for python < 3.12. ([`90e7202`](https://github.com/adfinis/document-merge-service/commit/90e72025076601d60aea27f3deac559a0c3e5770))

## 8.1.0 (22 October 2025)

### Feature

* **s3:** Add setting for CLIENT_CONFIG ([`658f1d7`](https://github.com/adfinis/document-merge-service/commit/658f1d7a3b324a51efc69473946f1b6e2337509f))
* **deps:** Update django to v5.2 LTS ([`f91a39f`](https://github.com/adfinis/document-merge-service/commit/f91a39f4a35f0b21fd360a3a860bea340893bb11))

## 8.0.3 (29 September 2025)

### Fix

* **docx-template:** Corruption of libreoffice-originated templates ([`d7b3cec`](https://github.com/adfinis-sygroup/document-merge-service/commit/d7b3cec727457e1f75d1e67d385c9c5d962cca64))

## 8.0.2 (25. July 2025)

### Feature

* **placeholders:** add specific error messages if a "-" is used in a placeholder ([e75252f](https://github.com/adfinis/document-merge-service/commit/e75252fc9d57c4732b23fde8e0143c5908d3bbf5))

### Fix

* **docs:** Correct typo ([58e2d05](https://github.com/adfinis/document-merge-service/commit/58e2d05eb081c397eb0c76494ca6ff7833544e8c))

## 8.0.0 (16. May 2025)

### Feature

* **docker:** Add image variants ([`72fb354`](https://github.com/adfinis/document-merge-service/commit/72fb354aef0f2b4dbd9d1249d2cc6d8df4f30aba))
* **docker:** Replace uwsgi with gunicorn as app server ([`aa3f24b`](https://github.com/adfinis/document-merge-service/commit/aa3f24b7f49c30a4b969ba83f89c522a4779929d))
* **build:** Use multi stage docker build for smaller images ([`73b9e43`](https://github.com/adfinis/document-merge-service/commit/73b9e4351636da53629acd4fe54eb9fba0c2855a))
* **docker:** Use python 3.13 for docker image ([`e38f4a4`](https://github.com/adfinis/document-merge-service/commit/e38f4a4fca21be43d93196ac1675ab223cd1bbdf))
* **deps:** Add python 3.13 to compatibility matrix ([`05e9268`](https://github.com/adfinis/document-merge-service/commit/05e9268dc28d209f9099ad2ea889b0b9b9f14cb9))
* **engine:** Remove obsolete docx mailmerge engine ([`698da36`](https://github.com/adfinis/document-merge-service/commit/698da3682b5657f3290ed274c4caf312c2e489f7))

### Fix

* **settings:** Make DOCXTEMPLATE_JINJA_EXTENSIONS setting optional ([`2ec2b63`](https://github.com/adfinis/document-merge-service/commit/2ec2b63b455792e4c586fc51f311f9cbc901d3f0))
* **deps:** Replace imghdr with python-magic ([`b6a4ca2`](https://github.com/adfinis/document-merge-service/commit/b6a4ca211a2aef19aefac7cc0bc4ce6998f86f0b))

### Breaking

* Remove dependency on MySQL. If you need to use mysql as a database, you may install the required dependency on top of the standard docker image. ([`08d909d`](https://github.com/adfinis/document-merge-service/commit/08d909db0adf7e32b0ed04fb4c00cc88c302be01))
* `document-merge-service` now uses Gunicorn as its app server instead of uWSGI. If you are using a custom uWSGI configuration via the `UWSGI_INI` env variable, replace it with a custom Gunicorn configuration as explained in [the documentation](CONFIGURATION.md#gunicorn). ([`aa3f24b`](https://github.com/adfinis/document-merge-service/commit/aa3f24b7f49c30a4b969ba83f89c522a4779929d))
* This will remove poetry entirely from the production image. If you customized the command, make sure to remove `poetry run` as the binaries are now globally available without using poetry. ([`73b9e43`](https://github.com/adfinis/document-merge-service/commit/73b9e4351636da53629acd4fe54eb9fba0c2855a))
* This fully removes the support of the docx mailmerge engine as the used library is quite inactive and there is no current use case for this engine. For more information, check issue #570. ([`698da36`](https://github.com/adfinis/document-merge-service/commit/698da3682b5657f3290ed274c4caf312c2e489f7))

## 7.3.7 (10 April 2025)

### Fix

* **ci:** Create container images also when pusblishing a draft release ([fb79d06](https://github.com/adfinis/document-merge-service/commit/fb79d06490661b06c66049b10d292b5ed92dcf55))

## 7.3.6 (8 April 2025)

### Fix

* **deps:** Fix locked dependency on django-storages ([`bf085a9`](https://github.com/adfinis/document-merge-service/commit/bf085a9808bdb76bb70bb8e834b92b6290384c73))
* **api:** Delete old file when updating template file ([`f98e38b`](https://github.com/adfinis/document-merge-service/commit/f98e38bdb1c1fff84e9e4f3c82d234250cd44c8d))

## 7.3.5 (13 January 2025)

### Fix

* **deps:** Update dependencies ([`16d4f76`](https://github.com/adfinis/document-merge-service/commit/16d4f767999989f46368f475e15a6335b5c06b28))

## 7.3.4 (12 December 2024)

### Fix

* **cache:** Cast common options for memcached cache ([`59f1727`](https://github.com/adfinis/document-merge-service/commit/59f1727b872f34da51757c894db5385170ff4fb4))

## 7.3.3 (10 December 2024)

### Fix

* **cache:** Add environment variable to configure cache options ([`6b45365`](https://github.com/adfinis/document-merge-service/commit/6b45365ac6cbc60b1739c75982f54934d57638bc))

## 7.3.2 (29 November 2024)

### Fix


document_merge_service/settings.py
import os
import re

import environ

from .sentry import sentry_init

env = environ.Env()
django_root = environ.Path(__file__) - 2

ENV_FILE = env.str("ENV_FILE", default=django_root(".env"))
if os.path.exists(ENV_FILE):  # pragma: no cover
    environ.Env.read_env(ENV_FILE)

# per default production is enabled for security reasons
# for development create .env file with ENV=development
ENV = env.str("ENV", "production")


def default(default_dev=env.NOTSET, default_prod=env.NOTSET):
    """Environment aware default."""
    return default_prod if ENV == "production" else default_dev


SECRET_KEY = env.str("SECRET_KEY", default=default("uuuuuuuuuu"))
DEBUG = env.bool("DEBUG", default=default(True, False))
ALLOWED_HOSTS = env.list("ALLOWED_HOSTS", default=default(["*"]))


# Database
# https://docs.djangoproject.com/en/1.11/ref/settings/#databases

DATABASE_DIR = env.str("DATABASE_DIR", default="/var/lib/document-merge-service/data")
DATABASES = {
    "default": {
        "ENGINE": env.str("DATABASE_ENGINE", default="django.db.backends.sqlite3"),
        "NAME": env.str("DATABASE_NAME", default=f"{DATABASE_DIR}/sqlite3.db"),
        "USER": env.str("DATABASE_USER", default=""),
        "PASSWORD": env.str("DATABASE_PASSWORD", default=""),
        "HOST": env.str("DATABASE_HOST", default=""),
        "PORT": env.str("DATABASE_PORT", default=""),
        "OPTIONS": env.dict("DATABASE_OPTIONS", default={}),
    }
}


# Application definition

INSTALLED_APPS = [
    "django.contrib.messages",
    "django.contrib.staticfiles",
    "rest_framework",
    "django_filters",
    "document_merge_service.api.apps.DefaultConfig",
    "corsheaders",
    "generic_permissions.apps.GenericPermissionsConfig",
]

if "postgresql" in DATABASES["default"]["ENGINE"]:  # pragma: no cover
    INSTALLED_APPS.append("django.contrib.postgres")

MIDDLEWARE = [
    "django.middleware.security.SecurityMiddleware",
    "corsheaders.middleware.CorsMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.locale.LocaleMiddleware",
    "whitenoise.middleware.WhiteNoiseMiddleware",
]

ROOT_URLCONF = "document_merge_service.urls"
WSGI_APPLICATION = "document_merge_service.wsgi.application"

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.messages.context_processors.messages",
            ]
        },
    }
]


# Cache
# https://docs.djangoproject.com/en/1.11/ref/settings/#caches

CACHES = {
    "default": {
        "BACKEND": env.str(
            "CACHE_BACKEND", default="django.core.cache.backends.locmem.LocMemCache"
        ),
        "LOCATION": env.str("CACHE_LOCATION", default=""),
        "OPTIONS": env.dict(
            "CACHE_OPTIONS",
            default={},
            cast={
                "value": str,
                # Common options for PyMemcacheCache which need to have the
                # proper datatype in order to work
                "cast": {
                    "connect_timeout": float,
                    "timeout": float,
                    "no_delay": bool,
                    "ignore_exc": bool,
                },
            },
        ),
    }
}


# CORS
CORS_ORIGIN_ALLOW_ALL = env.bool("CORS_ORIGIN_ALLOW_ALL", False)
CORS_ORIGIN_REGEX_WHITELIST = [r"^(https?://)?127\.0\.0\.1:\d{4}$"]
CORS_ORIGIN_REGEX_WHITELIST += env.list(
    "CORS_ORIGIN_REGEX_WHITELIST", default=[r"^(https?://)?127\.0\.0\.1:\d{4}$"]
)


# Internationalization
# https://docs.djangoproject.com/en/1.11/topics/i18n/

LANGUAGE_CODE = env.str("LANGUAGE_CODE", "en-us")
TIME_ZONE = env.str("TIME_ZONE", "UTC")
USE_I18N = True
USE_TZ = True


def parse_admins(admins):
    """
    Parse env admins to django admins.

    Example of ADMINS environment variable:
    Test Example <test@example.com>,Test2 <test2@example.com>
    """
    result = []
    for admin in admins:
        match = re.search(r"(.+) \<(.+@.+)\>", admin)
        if not match:  # pragma: no cover
            raise environ.ImproperlyConfigured(
                'In ADMINS admin "{0}" is not in correct '
                '"Firstname Lastname <email@example.com>"'.format(admin)
            )
        result.append((match.group(1), match.group(2)))
    return result


ADMINS = parse_admins(env.list("ADMINS", default=[]))


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/1.11/howto/static-files/

STATIC_URL = "/static/"
STATIC_ROOT = os.path.join(django_root, "staticfiles")

# Media files

STORAGES = {
    "default": {
        "BACKEND": env.str(
            "FILE_STORAGE", default="django.core.files.storage.FileSystemStorage"
        )
    },
    "staticfiles": {
        "BACKEND": "whitenoise.storage.CompressedManifestStaticFilesStorage",
    },
}
MEDIA_ROOT = env.str("MEDIA_ROOT", "")
# TODO: This should be removed in favor of storing the files in a bucket
# https://code.djangoproject.com/ticket/32991
MEDIA_URL = env.str("MEDIA_URL", "api/v1/template/")

# django-storages S3 settings
DMS_ENABLE_AT_REST_ENCRYPTION = env.bool("DMS_ENABLE_AT_REST_ENCRYPTION", False)
S3_STORAGE_OPTIONS = {
    "access_key": env.str("DMS_S3_ACCESS_KEY_ID", "minio"),
    "secret_key": env.str("DMS_S3_SECRET_ACCESS_KEY", "minio123"),
    "bucket_name": env.str("DMS_STORAGE_BUCKET_NAME", "dms-media"),
    "endpoint_url": env.str("DMS_S3_ENDPOINT_URL", "http://minio:9000"),
    "region_name": env.str("DMS_S3_REGION_NAME", None),
    "location": env.str("DMS_LOCATION", ""),
    "file_overwrite": env.bool("DMS_S3_FILE_OVERWRITE", False),
    "signature_version": env.str("DMS_S3_SIGNATURE_VERSION", "v2"),
    "use_ssl": env.bool("DMS_S3_USE_SSL", default=True),
    "verify": env.bool("DMS_S3_VERIFY", default=None),
}
if (
    STORAGES["default"]["BACKEND"] == "storages.backends.s3.S3Storage"
):  # pragma: no cover
    from botocore.config import Config as BotoConfig

    S3_STORAGE_OPTIONS["client_config"] = BotoConfig(
        **env.dict("DMS_S3_CLIENT_CONFIG", default={})
    )

    if DMS_ENABLE_AT_REST_ENCRYPTION:
        S3_STORAGE_OPTIONS["object_parameters"] = {
            "SSECustomerKey": env.str(
                "DMS_S3_STORAGE_SSEC_SECRET",
                default=default("x" * 32),
            ),
            "SSECustomerAlgorithm": "AES256",
        }
    STORAGES["default"]["OPTIONS"] = S3_STORAGE_OPTIONS

# Gotenberg
GOTENBERG_HOST = env.str("GOTENBERG_HOST", default="http://gotenberg")
GOTENBERG_PORT = env.int("GOTENBERG_PORT", default=3000)
GOTENBERG_URL = f"{GOTENBERG_HOST}:{GOTENBERG_PORT}"

GOTENBERG_PDF_METADATA = env.dict("GOTENBERG_PDF_METADATA", default={})
GOTENBERG_PDF_A_FORMAT = env.str("GOTENBERG_PDF_A_FORMAT", default=None)
GOTENBERG_PDF_UA = env.bool("GOTENBERG_PDF_UA", default=False)
GOTENBERG_PDF_FLATTEN = env.bool("GOTENBERG_PDF_FLATTEN", default=False)


# Jinja2
DOCXTEMPLATE_JINJA_EXTENSIONS = env.list("DOCXTEMPLATE_JINJA_EXTENSIONS", default=[])

# Authentication

REQUIRE_AUTHENTICATION = env.bool("REQUIRE_AUTHENTICATION", False)

OIDC_USERINFO_ENDPOINT = env.str("OIDC_USERINFO_ENDPOINT", default=None)
OIDC_VERIFY_SSL = env.bool("OIDC_VERIFY_SSL", default=True)
OIDC_GROUPS_CLAIM = env.str("OIDC_GROUPS_CLAIM", default="")
OIDC_USERNAME_CLAIM = env.str("OIDC_USERNAME_CLAIM", default="sub")
OIDC_BEARER_TOKEN_REVALIDATION_TIME = env.int(
    "OIDC_BEARER_TOKEN_REVALIDATION_TIME", default=0
)

# Rest framework
# https://www.django-rest-framework.org/api-guide/settings/

REST_FRAMEWORK = {
    "DEFAULT_PERMISSION_CLASSES": [
        "document_merge_service.api.permissions.AsConfigured",
    ],
    "DEFAULT_AUTHENTICATION_CLASSES": [
        "document_merge_service.api.authentication.BearerTokenAuthentication"
    ],
    "UNAUTHENTICATED_USER": "document_merge_service.api.authentication.AnonymousUser",
    "DEFAULT_FILTER_BACKENDS": (
        "rest_framework.filters.OrderingFilter",
        "django_filters.rest_framework.DjangoFilterBackend",
        "rest_framework.filters.SearchFilter",
    ),
    "TEST_REQUEST_DEFAULT_FORMAT": "json",
}

PAGINATION_ENABLED = env.bool("PAGINATION_ENABLED", True)
PAGINATION_DEFAULT_PAGE_SIZE = env.int("PAGINATION_DEFAULT_PAGE_SIZE", 100)
PAGINATION_MAX_PAGE_SIZE = env.int("PAGINATION_MAX_PAGE_SIZE", 1000)

if PAGINATION_ENABLED:
    REST_FRAMEWORK.update(
        {
            "DEFAULT_PAGINATION_CLASS": "document_merge_service.api.pagination.APIPagination",
            "PAGE_SIZE": PAGINATION_DEFAULT_PAGE_SIZE,
        }
    )

# Logging
ENABLE_ADMIN_EMAIL_LOGGING = env.bool("ENABLE_ADMIN_EMAIL_LOGGING", False)

LOGGING = {
    "version": 1,
    "disable_existing_loggers": False,
    "handlers": {
        "console": {
            "level": "WARNING",
            "filters": None,
            "class": "logging.StreamHandler",
        },
        "mail_admins": {
            "level": "ERROR",
            "filters": None,
            "class": "django.utils.log.AdminEmailHandler",
        },
    },
    "loggers": {"django": {"handlers": ["console"], "level": "WARNING"}},
}

URL_PREFIX = env.str("URL_PREFIX", default="")

# Email settings
SERVER_EMAIL = env.str("SERVER_EMAIL", default="root@localhost")
DEFAULT_FROM_EMAIL = env.str("DEFAULT_FROM_EMAIL", default="webmaster@localhost")
EMAIL_HOST = env.str("EMAIL_HOST", default="localhost")
EMAIL_PORT = env.int("EMAIL_PORT", default=25)
EMAIL_HOST_USER = env.str("EMAIL_HOST_USER", default="")
EMAIL_HOST_PASSWORD = env.str("EMAIL_HOST_PASSWORD", default="")
EMAIL_USE_TLS = env.bool("EMAIL_USE_TLS", default=False)

# Email error handler
if ENABLE_ADMIN_EMAIL_LOGGING:  # pragma: no cover
    LOGGING["loggers"]["django"]["handlers"].append("mail_admins")  # type: ignore

# Sentry error tracking
SENTRY_DSN = env.str("SENTRY_DSN", default="")
SENTRY_ENVIRONMENT = env.str("SENTRY_ENVIRONMENT", default="development")
SENTRY_TRACES_SAMPLE_RATE = env.float("SENTRY_TRACES_SAMPLE_RATE", default=0.01)
SENTRY_SEND_DEFAULT_PII = env.bool("SENTRY_SEND_DEFAULT_PII", default=False)

if SENTRY_DSN:  # pragma: no cover
    sentry_init(
        SENTRY_DSN,
        SENTRY_ENVIRONMENT,
        SENTRY_TRACES_SAMPLE_RATE,
        SENTRY_SEND_DEFAULT_PII,
    )

# https://github.com/adfinis/django-generic-api-permissions
GENERIC_PERMISSIONS_PERMISSION_CLASSES = env.list("DMS_PERMISSION_CLASSES", default=[])
GENERIC_PERMISSIONS_VISIBILITY_CLASSES = env.list("DMS_VISIBILITY_CLASSES", default=[])

# App specific arguments for the extension classes
EXTENSIONS_ARGUMENTS = env.dict("EXTENSIONS_ARGUMENTS", default={})

# DMS potentially uses a large number of fields when creating templates due to
# the possibility to validate a template against available placeholders.
# https://docs.djangoproject.com/en/5.1/ref/settings/#data-upload-max-number-fields
DATA_UPLOAD_MAX_NUMBER_FIELDS = env.int("DATA_UPLOAD_MAX_NUMBER_FIELDS", default=1000)
